File: app/src/main/java/com/sjani/usnationalparkguide/UI/Details/Trails/TrailDetailViewModel.java
package com.sjani.usnationalparkguide.UI.Details.Trails;


import com.sjani.usnationalparkguide.Data.TrailEntity;
import com.sjani.usnationalparkguide.Utils.ParkRepository;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

public class TrailDetailViewModel extends ViewModel {

    private final ParkRepository repository;
    private String trailApiKey;
    private String latLong;
    private LiveData<TrailEntity> trail;

    public TrailDetailViewModel(ParkRepository repository, String trailApiKey, String latLong) {
        this.repository = repository;
        this.trailApiKey = trailApiKey;
        this.latLong = latLong;

    }

    public LiveData<TrailEntity> getTrail(String trailId) {
        trail = this.repository.getTrailFromDb(trailId);
        return trail;
    }


}

File: app/src/main/java/com/sjani/usnationalparkguide/ui/viewmodel/TrailDetailViewModel.kt
package com.sjani.usnationalparkguide.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.sjani.usnationalparkguide.data.entity.TrailEntity
import com.sjani.usnationalparkguide.data.repository.ParkRepository
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn

class TrailDetailViewModel(
    repository: ParkRepository,
    trailId: String
) : ViewModel() {
    
    val trail = repository.getTrail(trailId)
        .stateIn(viewModelScope, SharingStarted.Lazily, null)
}

class TrailDetailViewModelFactory(
    private val repository: ParkRepository,
    private val trailId: String
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(TrailDetailViewModel::class.java)) {
            return TrailDetailViewModel(repository, trailId) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}


